src/sdam/server_selection_events.ts

```diff
@@ -0,0 +1,137 @@
+import {
+  SERVER_SELECTION_FAILED,
+  SERVER_SELECTION_STARTED,
+  SERVER_SELECTION_SUCCEEDED,
+  WAITING_FOR_SUITABLE_SERVER
+} from '../constants';
+import { type ReadPreference } from '../read_preference';
+import { type ServerSelector } from './server_selection';
+import type { TopologyDescription } from './topology_description';
+
+/**
+ * The base export class for all logs published from server selection
+ * @internal
+ * @category Log Type
+ */
+export abstract class ServerSelectionEvent {
+  /** String representation of the selector being used to select the server.
+   *  Defaults to 'custom selector' for application-provided custom selector case.
+   */
+  selector: string | ReadPreference | ServerSelector;
+  /** The name of the operation for which a server is being selected.  */
+  operation: string;
+  /** 	The current topology description.  */
+  topologyDescription: TopologyDescription;
+
+  /** @internal */
+  abstract name:
+    | typeof SERVER_SELECTION_STARTED
+    | typeof SERVER_SELECTION_SUCCEEDED
+    | typeof SERVER_SELECTION_FAILED
+    | typeof WAITING_FOR_SUITABLE_SERVER;
+
+  abstract message: string;
+
+  /** @internal */
+  constructor(
+    selector: string | ReadPreference | ServerSelector,
+    topologyDescription: TopologyDescription,
+    operation?: string
+  ) {
+    this.selector = selector;
+    this.operation = operation ?? 'n/a';
+    this.topologyDescription = topologyDescription;
+  }
+}
+
+/**
+ * An event published when server selection starts
+ * @internal
+ * @category Event
+ */
+export class ServerSelectionStartedEvent extends ServerSelectionEvent {
+  /** @internal */
+  name = SERVER_SELECTION_STARTED;
+  message = 'Server selection started';
+
+  /** @internal */
+  constructor(
+    selector: string | ReadPreference | ServerSelector,
+    topologyDescription: TopologyDescription,
+    operation?: string
+  ) {
+    super(selector, topologyDescription, operation);
+  }
+}
+
+/**
+ * An event published when a server selection fails
+ * @internal
+ * @category Event
+ */
+export class ServerSelectionFailedEvent extends ServerSelectionEvent {
+  /** @internal */
+  name = SERVER_SELECTION_FAILED;
+  message = 'Server selection failed';
+  /** Representation of the error the driver will throw regarding server selection failing. */
+  failure: Error;
+
+  /** @internal */
+  constructor(
+    selector: string | ReadPreference | ServerSelector,
+    topologyDescription: TopologyDescription,
+    error: Error,
+    operation?: string
+  ) {
+    super(selector, topologyDescription, operation);
+    this.failure = error;
+  }
+}
+
+/**
+ * An event published when server selection succeeds
+ * @internal
+ * @category Event
+ */
+export class ServerSelectionSucceededEvent extends ServerSelectionEvent {
+  /** @internal */
+  name = SERVER_SELECTION_SUCCEEDED;
+  message = 'Server selection succeeded';
+  /** The address (host/port pair) of the pool */
+  address: string;
```

Comment:
The spec says `This message MUST contain the following key-value pairs: message, serverHost , serverPort `, so maybe we shout split address to those?

Reply:
Good idea! I've added in these changes.